ch55xtool/ch55xtool.py

```diff
@@ -4,426 +4,916 @@
-            payload[index] = (
-                payload[index] ^ (
-                    (chk_sum + chip_id) % 256)) % 256
-
-    left_len = file_length
-    curr_addr = 0
-
-    while curr_addr < file_length:
-
-        if left_len >= 56:
-            pkt_length = 56
-        else:
-            pkt_length = left_len
-
-        __VERIFY_CMD_V2 = VERIFY_CMD_V2
-        __VERIFY_CMD_V2[1] = pkt_length + 5
-        __VERIFY_CMD_V2[3] = curr_addr % 256
-        __VERIFY_CMD_V2[4] = (curr_addr >> 8) % 256
-        __VERIFY_CMD_V2[7] = left_len % 256
-        __VERIFY_CMD_V2 = __VERIFY_CMD_V2 + \
-            payload[curr_addr:curr_addr + pkt_length]
-
-        dev.write(EP_OUT_ADDR, __VERIFY_CMD_V2)
-        ret = dev.read(EP_IN_ADDR, 6, USB_MAX_TIMEOUT)
-
-        curr_addr = curr_addr + pkt_length
-        left_len = left_len - pkt_length
-
-        if ret[4] != 0x00:
-            return None
-
-    return file_length
-
-
-def __verify_flash_ch55x_v23(dev, chk_sum, chip_id, payload):
-    # Payload needs to be padded, 56 is a good number
-    payload = payload + [0] * \
-        ((math.ceil(len(payload) / 56) * 56) - len(payload))
-    file_length = len(payload)
-
-    for index in range(file_length):
-        if index % 8 == 7:
-            payload[index] = (
-                payload[index] ^ (
-                    (chk_sum + chip_id) % 256)) % 256
-        else:
-            payload[index] = (payload[index] ^ chk_sum) % 256
-
-    left_len = file_length
-    curr_addr = 0
-
-    while curr_addr < file_length:
-
-        if left_len >= 56:
-            pkt_length = 56
-        else:
-            pkt_length = left_len
-
-        __VERIFY_CMD_V2 = VERIFY_CMD_V2
-        __VERIFY_CMD_V2[1] = pkt_length + 5
-        __VERIFY_CMD_V2[3] = curr_addr % 256
-        __VERIFY_CMD_V2[4] = (curr_addr >> 8) % 256
-        __VERIFY_CMD_V2[7] = left_len % 256
-        __VERIFY_CMD_V2 = __VERIFY_CMD_V2 + \
-            payload[curr_addr:curr_addr + pkt_length]
-
-        dev.write(EP_OUT_ADDR, __VERIFY_CMD_V2)
-        ret = dev.read(EP_IN_ADDR, 6, USB_MAX_TIMEOUT)
-
-        curr_addr = curr_addr + pkt_length
-        left_len = left_len - pkt_length
-
-        if ret[4] != 0x00:
-            return None
-
-    return file_length
-
-
-def __end_flash_ch55x_v2(dev):
-    dev.write(EP_OUT_ADDR, END_FLASH_CMD_V2)
-    ret = dev.read(EP_IN_ADDR, 6, USB_MAX_TIMEOUT)
-    if ret[4] != 0x00:
-        return None
-    else:
-        return True
+USB_MAX_TIMEOUT = 5000
 
+# =============================================
+WCH_CMDS = 	{	"Detect":		b'\xA1',
+				"End":			b'\xA2',
+				"SetKey":		b'\xA3',
+				"FlashErase":	b'\xA4',
+				"FlashWrite":	b'\xA5',
+				"FlashVerify":	b'\xA6',
+				"ReadConfig":	b'\xA7',
+				"WriteConfig":	b'\xA8',
+				"DataErase":	b'\xA9',
+				"DataWrite":	b'\xAA',
+				"DataRead":		b'\xAB',
+				"ReadOTP":		b'\xC4',
+				"WriteOTP":		b'\xC3',
+			}
+
+# Meaning of those two values not yet clear :(
+DETECT_PL_START = b'\x42\x10'
```

Comment:
first is `ChipId`
the second is `DeviceType`

Reply:
Thanks, yes, it's known now and it starts from 0x10 value (as reported by Patrick Yang from WCH).
It is DeviceType = ChipSeries = McuType (+0x10 from typeall.wcfg)
I plan to implement chip selecting dependency on it too near time.

Reply:
I tried to switch boards and detect another chip and detection change those values in some order.
I switch from CH568 to CH552 and detection still was working even with wrong values.

Reply:
And still a bit unclear why:
for CH552 it is 0x52
but
for CH568 it is 0x42 (not 0x68)

Reply:
My test has the following conclusions:

All CH56x family chips have chip_id = 0x42. when you see a 0x42, it must be one of `CH56x`.

Reply:
Which chip you work with ? Is it CH563 ?
From reverse engineering of WCHISPTool I have:
```
      switch ( ... )
      {
        case 0:
        case 0x46:
        case 0x61:
          LOBYTE(dword_100240BC) = 70;
          strcpy(Str, aCh561);
          break;
        case 1:
        case 0x42:
        case 0x63:
          LOBYTE(dword_100240BC) = 66;
          strcpy(Str, aCh563);
          break;
        case 2:
        case 0x65:
          LOBYTE(dword_100240BC) = 101;
          strcpy(Str, aCh565);
          break;
        case 3:
        case 0x66:
          LOBYTE(dword_100240BC) = 102;
          strcpy(Str, aCh566);
          break;
        case 4:
        case 0x67:
          LOBYTE(dword_100240BC) = 103;
          strcpy(Str, aCh567);
          break;
        case 5:
        case 0x68:
          LOBYTE(dword_100240BC) = 104;
          strcpy(Str, aCh568);
          break;
        case 6:
        case 0x69:
          LOBYTE(dword_100240BC) = 105;
          strcpy(Str, aCh569);
          break;
        default:
          break;
      }
```
and CH568 I have report it ID as 0x68 , but yes, detection request is going with 0x42 .
And it response with correct id to 0x68 request too.

Reply:
I finally built a WCH chip DB from all available information, embedded in my tool:

https://github.com/ch32-rs/wchisp/tree/main/devices

Reply:
It's huge positive push for me to migrate to those configs :)
Never worked with YAML before but looks like it's time start.

Reply:
Hope it helps.
now it can recognize my CH32V307 chip's variants correctly. (flash size matters for erase command).